pkg/agent/controller/bgp/controller.go

```diff
@@ -0,0 +1,954 @@
+		k8sClient:                 k8sClient,
+		bgpPeerPasswords:          make(map[string]string),
+		nodeName:                  nodeConfig.Name,
+		enabledIPv4:               networkConfig.IPv4Enabled,
+		enabledIPv6:               networkConfig.IPv6Enabled,
+		podIPv4CIDR:               nodeConfig.PodIPv4CIDR.String(),
+		podIPv6CIDR:               nodeConfig.PodIPv6CIDR.String(),
+		nodeIPv4Addr:              nodeConfig.NodeIPv4Addr.IP.String(),
+		egressEnabled:             features.DefaultFeatureGate.Enabled(features.Egress),
+		newBGPServerFn: func(globalConfig *bgp.GlobalConfig) bgp.Interface {
+			return gobgp.NewGoBGPServer(globalConfig)
+		},
+		queue: workqueue.NewNamedRateLimitingQueue(workqueue.NewItemExponentialFailureRateLimiter(minRetryDelay, maxRetryDelay), "bgpPolicyGroup"),
+	}
+	c.bgpPolicyInformer.AddEventHandlerWithResyncPeriod(
+		cache.ResourceEventHandlerFuncs{
+			AddFunc:    c.addBGPPolicy,
+			UpdateFunc: c.updateBGPPolicy,
+			DeleteFunc: c.deleteBGPPolicy,
+		},
+		resyncPeriod,
+	)
+	c.serviceInformer.AddEventHandlerWithResyncPeriod(
+		cache.ResourceEventHandlerFuncs{
+			AddFunc:    c.addService,
+			UpdateFunc: c.updateService,
+			DeleteFunc: c.deleteService,
+		},
+		resyncPeriod,
+	)
+	c.endpointSliceInformer.AddEventHandlerWithResyncPeriod(
+		cache.ResourceEventHandlerFuncs{
+			AddFunc:    c.addEndpointSlice,
+			UpdateFunc: c.updateEndpointSlice,
+			DeleteFunc: nil,
+		},
+		resyncPeriod,
+	)
+	if c.egressEnabled {
+		c.egressInformer.AddEventHandlerWithResyncPeriod(
+			cache.ResourceEventHandlerFuncs{
+				AddFunc:    c.addEgress,
+				UpdateFunc: c.updateEgress,
+				DeleteFunc: c.deleteEgress,
+			},
+			resyncPeriod,
+		)
+	}
+	c.nodeInformer.AddEventHandlerWithResyncPeriod(
+		cache.ResourceEventHandlerFuncs{
+			AddFunc:    nil,
+			UpdateFunc: c.updateNode,
+			DeleteFunc: nil,
+		},
+		resyncPeriod,
+	)
+	return c, nil
+}
+
+// watchSecretChanges uses watch API directly to watch for the changes of the specific Secret storing passwords of BGP
+// peers.
+func (c *Controller) watchSecretChanges(endCh <-chan struct{}) error {
+	ns := env.GetAntreaNamespace()
+	watcher, err := c.k8sClient.CoreV1().Secrets(ns).Watch(context.TODO(), metav1.SingleObject(metav1.ObjectMeta{
+		Namespace: ns,
+		Name:      types.BGPPolicySecretName,
+	}))
+	if err != nil {
+		return fmt.Errorf("failed to create Secret watcher: %v", err)
+	}
+
+	ch := watcher.ResultChan()
+	defer watcher.Stop()
+	klog.InfoS("Starting watching Secret changes", "Secret", fmt.Sprintf("%s/%s", ns, types.BGPPolicySecretName))
+	for {
+		select {
+		case event, ok := <-ch:
+			if !ok {
+				return nil
+			}
+			// Update BGP peer passwords.
+			klog.InfoS("Processing Secret event", "Secret", fmt.Sprintf("%s/%s", ns, types.BGPPolicySecretName))
+			func() {
+				c.bgpPeerPasswordsMutex.Lock()
+				defer c.bgpPeerPasswordsMutex.Unlock()
+
+				secretObj := event.Object.(*corev1.Secret)
```

Comment:
The event may be Deleted or Bookmark, in which case it's not correct to access its Data?

Reply:
My bad. I should have thought of that. Updated as you said.